src/Storages/MergeTree/ParallelReplicasReadingCoordinator.cpp

```diff
@@ -1087,17 +1116,60 @@ ParallelReadResponse ParallelReplicasReadingCoordinator::handleRequest(ParallelR
 
     ProfileEventTimeIncrement<Microseconds> watch(ProfileEvents::ParallelReplicasHandleRequestMicroseconds);
 
-    std::lock_guard lock(mutex);
+    ParallelReadResponse response;
+    response.finish = true;
 
-    if (!pimpl)
-        initialize(request.mode);
+    if (is_reading_completed)
+        return response;
 
-    const auto replica_num = request.replica_num;
-    auto response = pimpl->handleRequest(std::move(request));
-    if (!response.finish)
+    std::set<size_t> replicas_to_exclude;
+    bool reading_assignment_has_been_completed = false;
     {
-        if (replicas_used.insert(replica_num).second)
-            ProfileEvents::increment(ProfileEvents::ParallelReplicasUsedCount);
+        std::lock_guard lock(mutex);
+        if (is_reading_completed)
+            return response;
+
+        if (!pimpl)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Got read request from replica {} without ranges announcement", request.replica_num);
+
+        const auto replica_num = request.replica_num;
+        response = pimpl->handleRequest(std::move(request));
+        if (!response.finish)
+        {
+            chassert(!is_reading_completed);
+
+            if (replicas_used.insert(replica_num).second)
+                ProfileEvents::increment(ProfileEvents::ParallelReplicasUsedCount);
+        }
+        else
+        {
+            if (isReadingCompleted())
```

Comment:
maybe makes sense to check if `is_reading_completed` is already `true` to avoid calling `isReadingCompleted()`.

Reply:
It's done on line 1122